### DISCUSSION TITLE
List of frequencies in mp3, and amplitude of frequency vs timeline

### DISCUSSION THREAD

**Discussion by kishorpise** [naudio/NAudio#1174] [Q&A]
Hello,
I am looking for function- to list  all frequencies in mp3 file. some how its working but giving avg amplitude.  but not able to get the amplitude on time line. can some one help ? C# asp.net core. here is code sample.

So my function ideally should take frequency as param- and output I should get the time series with amplitude. is this possible, windowing could be 5ms or 1 sec  ?

Any help is appriciated.

```

  public List<Tuple<double, double>> GetFrequencies(string mp3FilePath)
  {
      List<Tuple<double, double>> frequencies = new List<Tuple<double, double>>();

      using (var reader = new AudioFileReader(mp3FilePath))
      {
          int sampleRate = reader.WaveFormat.SampleRate;
          int fftLength = 1024; // Length of FFT window

          // Create buffers for audio data and FFT
          float[] audioBuffer = new float[fftLength];
          Complex[] fftBuffer = new Complex[fftLength];

          // Prepare FFT
          int fftSamples = fftLength / 2;
          double[] fftReal = new double[fftSamples];
          double[] fftImaginary = new double[fftSamples];

          int samplesRead;
          while ((samplesRead = reader.Read(audioBuffer, 0, fftLength)) > 0)
          {
              // Convert audio data to complex numbers for FFT
              for (int i = 0; i < fftLength; i++)
              {
                  fftBuffer[i] = new Complex(audioBuffer[i], 0);
              }

              // Perform FFT
              Fourier.Forward(fftBuffer);

              // Calculate magnitude spectrum
              double[] magnitude = new double[fftSamples];
              for (int i = 0; i < fftSamples; i++)
              {
                  magnitude[i] = fftBuffer[i].Magnitude;
              }

              // Find peak magnitude and corresponding frequency
              double maxMagnitude = magnitude[0];
              int maxIndex = 0;
              for (int i = 1; i < fftSamples; i++)
              {
                  if (magnitude[i] > maxMagnitude)
                  {
                      maxMagnitude = magnitude[i];
                      maxIndex = i;
                  }
              }

              // Calculate frequency corresponding to maxIndex
              double frequency = (double)maxIndex * sampleRate / fftLength;
              frequencies.Add(new Tuple<double, double>(frequency, maxMagnitude));
          }
      }

      return frequencies;
  }


Here is sample output 

```
[
  {
    "item1": 0,
    "item2": 0
  },  
  {
    "item1": 93.75,
    "item2": 0.07660996136293814
  },
  {
    "item1": 93.75,
    "item2": 2.85952691449515
  },
  {
    "item1": 93.75,
    "item2": 2.6967089407631897
  },
  {
    "item1": 93.75,
    "item2": 3.9119828431779307
  },
  {
    "item1": 93.75,
    "item2": 2.3156258142369417
  },
  {
    "item1": 234.375,
    "item2": 1.572490876906007
  },

**Comment by markheath** [comment]
Your code is finding the peak bin per window. For "amplitude of frequency X over time" you instead want to read the magnitude of the specific bin(s) for X out of every window. The bin for a frequency is `f * fftLength / sampleRate`, and each FFT window is one point on your timeline (`time = windowIndex * fftLength / sampleRate`). 

Smaller windows = finer time resolution but coarser frequency resolution — 1024 samples at 44.1kHz is ~23ms/window and ~43Hz/bin, so a 5ms window can't resolve low frequencies. 

If you only care about one frequency, a Goertzel filter is much cheaper than a full FFT but NAudio does not include one.
